### ISSUE TITLE
IE11 "Unable to get property 'bind' of undefined or null reference - main.js (5,9402)" when using Promise ponyfill

### ISSUE THREAD

**Issue by tomknappramos** [preactjs/preact#2345]
Setup: 
 - Webpack + Babel.
 - I can't touch the global scope, so I can't polyfill. 
 - I'm using _Webpack Provide Plugin_ + _native-promise-only-ponyfill_ for Promise support.

My code:
```
import { h, render, Component } from 'preact';
```

When I clicked the error, I found this in the Preact build:
```
e.prototype.then.bind(e.resolve())
```

It appears as though "e" is the minified name for "Promise", and that the error is coming from around there in the code. 

Is Preact intentionally touching the prototype of Promise? Would there be a reason for this? Or is this an artifact of a build tool?

**Comment by tomknappramos** [comment]
PS
I just had the idea to try an older version of Preact, and version 8 of Preact didn't have the error.

**Comment by marvinhagemeister** [comment]
That sounds like the `Promise` polyfill your using isn't fully spec compliant. We use a promise tick as your schedule to process the render queue. If the `Promise` object isn't available we normally fallback to using `setTimeout` but since the polyfill is loaded it won't be used.

**Comment by tomknappramos** [comment]
Yeah, the ponyfill most likely isn't perfect, like all ponyfills/polyfills haha. 

However, because Preact is so lightweight and DOM-centric, it seems like Preact could potentially be a solid go-to framework for building **website plugins** (apps that exist within someone else's website and therefore must operate in isolation from the rest of the page which can't touch the global window, must work in basically all browsers (including beloved IE), and definitely not mixin/muddle with prototypes of native classes, etc).

Just curious, though, is there a reason why Preact is accessing the Promise prototype when implementing the Promise Tick? Or, is that just an artifact of a a build tool?

Seems like,
```
e.prototype.then.bind(e.resolve())
```

Could just be,
```
e.then.bind(e.resolve())
```

Right?

**Comment by marvinhagemeister** [comment]
Yup, the "islands" or widget approach as we call it is one of the main use cases for Preact. Your proposed solution doesn't work as the `then()` function only exists on the prototype:

```bash
> Promise.prototype.then
[Function: then]

> Promise.then
undefined
```

**Comment by marvinhagemeister** [comment]
After a bit more thinking, I don't think there is anything we can do on Preact's end here. Using incorrect polyfills that are not spec compliant is always a dangerous road and can lead to weird errors. It's not something we officially support in Preact.

The solution to fix this error is to use a spec compliant Promise polyfill that has the `Promise.prototype.then` method.

**Comment by tomknappramos** [comment]
Yeah, agreed. The Promise ponyfill doesn't have a **then** function in its prototype, so it's clear that this the ponyfill's fault (assuming the Promise spec is to have that implemented, which is likely).
 
Anyway, I appreciate the replies! I'll stick to v8 for now because it's happy without Promise.prototype.then and look around for a more spec-compliant ponyfill later. 

Lastly, aside from the error, just from a software design perspective, doing a **bind()** on the Promise class's prototype **then()** function, and then passing in **Promsie.resolve()** as the context is something I've never seen before.

Logically, that would return a copy of **Promise.prototype.then**  with the a resolved promise instance (**Promise.resolve()**) as the context. 
(note to readers: this does not modify the Promise class itself as bind returns a copy and leaves the original function untouched, so rest assured that your website widget/plugin is not messing with the global Promise class 👍)

Just for educational purposes, do you mind giving a quick explanation behind the purpose/usage behind that for readers of this issue?

Thanks!

**Comment by marvinhagemeister** [comment]
In Preact we use that trick to process our global queue in "ticks". The queue holds components that need to be rendered or otherwise updated. There are 3 ways components are pushed into the queue:

1. `component.setState()`
2. `component.forceUpdate()`
3. `render(<App/>, container)`

Note that even our `hooks` implementation makes use of classes internally and will call the same `setState` or `forceUpdate` function on the backing class of a component.

Once the queue has some items in it, we need to start processing it. But we don't want to do that immediately as the current diff should finish first, so that all dirty components of the same "commit" are processed at the same time. So we need to somehow schedule the invocation of when we should start processing it.

An easy way to do that in all browsers is via `setTimeout`. It fullfills our requirement to defer processing the queue until the current synchronous diff is completed.

```js
// Warning: Pseudo code!
const queue = [A, B, C]

function process() {
  let item;
  while ((item = queue.pop()) {
    queue.length
  }
}

function enqueueRender(component) {
  queue.push(component)
  // TODO: Missing here is a way to check if the processing
  // is already running
  process();
}

// Somewhere called in setState or forceUpdate
enqueueRender(currentComponent)
```

Now on more modern browser we do have more options on how we can schedule work. One of those is via so called [microtasks](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/). It's quicker and in our benchmarking tests more performant compared to `setTimeout` on modern browsers. To do that we have to leverage the `Promise` object. In a similar fashion to `setTimeout` we need to have a function that works the same way, but uses `Promises` under the hood.

Calling just `Promise.prototype.then()` directly **will** throw as a Promise expects to act on a receiving Promise that is passed via the `this` argument. To check that one can fire up node and paste the following lines:

```bash
> Promise.prototype.then(() => console.log("foo"))
Thrown:
TypeError: Method Promise.prototype.then called on incompatible receiver #<Promise>
    at Promise.then (<anonymous>)
> Promise.prototype.then.bind(Promise.resolve())(() => console.log("foo"))
Promise { <pending> }
> foo
```

As you can see above, calling `Promise.prototype.then` directly throws an error. Therefore we need to bind some Promise object to the `this` value of the `.then` method. We can do that via the `.bind()` operator, which brings us to the code that we use in Preact today:

```js
// Function to schedule the queue processing
const defer =
	typeof Promise == 'function'
		? Promise.prototype.then.bind(Promise.resolve())
		: setTimeout;
```

So in conclusion the main reason we use `Promises` instead of `setTimeout` to process our queue is performance. Hope that answers your questions :+1:

**Comment by jacobbogers** [comment]
Not so much a fan of "bind"-ing things
```javascript 
const defer = fn => Promise.resolve().then(fn);
```

**Comment by just-boris** [comment]
+1 to @jacobbogers point. Also that implementation is 4 characters shorter:

```js
Promise.prototype.then.bind(Promise.resolve()) // 46 chars
```

```js
function(fn){Promise.resolve().then(fn)} // 40 chars
```

**Comment by marvinhagemeister** [comment]
@jacobbogers @just-boris That's a great suggestion and it is indeed a few bytes shorter than what we have now :+1: But we have to think about the big picture here. This piece of code is somewhat in the "hot path" as it's called very frequently during the life of a Preact application. It's not as hot as `createElement/h`, but still it will be called quite a lot. Because of that any changes will affect benchmarks here.

```js
// Assumption: One time allocation
const defer = Promise.prototype.then.bind(Promise.resolve());

// Assumption: Allocates a new Promise whenever it is called
const defer = fn => Promise.resolve().then(fn);
```

An additional allocation here will affect benchmarks negatively like this one: https://github.com/krausest/js-framework-benchmark . One can of course rightfully argue about how meaningful those numbers are, but many developers still make decisions based on those numbers. The `.bind` approach seems to armortize that cost for us and the few benchmarks we have in our repo seem to back that up. There the `bind()` approach is slightly faster on my machine. So we trade a few bytes for runtime performance here.

Optimizations change frequently in browsers, so this may not be this way for forever. Just to be sure I posted the [same question on Twitter](https://mobile.twitter.com/marvinhagemeist/status/1229111485319827463) to some v8 engineers.

**Comment by jacobbogers** [comment]
understood about the hot path, this is a create the promise once not on ever call to defer
```
const defer = () =>function(){ var o = Promise.resolve(); return c => o.then(c) }();
````

**Comment by JoviDeCroock** [comment]
@jacobbogers yes but that will add a lot of bytes eventually so that wouldn't really be worth it

**Comment by jacobbogers** [comment]
Prolly not, using the bind is very straightforward, agree
